couchbase/src/main/scala/akka/stream/alpakka/couchbase/impl/CouchbaseCollectionSessionImpl.scala

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) since 2016 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.couchbase.impl
+
+import akka.annotation.InternalApi
+import akka.stream.alpakka.couchbase.CouchbaseDocument
+import akka.stream.alpakka.couchbase.scaladsl.{CouchbaseCollectionSession, CouchbaseSession}
+import akka.stream.scaladsl.Source
+import akka.{Done, NotUsed}
+import com.couchbase.client.java.codec.{RawBinaryTranscoder, RawStringTranscoder, Transcoder}
+import com.couchbase.client.java.json.JsonValue
+import com.couchbase.client.java.kv._
+import com.couchbase.client.java.manager.query.{CreateQueryIndexOptions, QueryIndex}
+import com.couchbase.client.java.{AsyncCollection, AsyncScope}
+import rx.{Observable, RxReactiveStreams}
+
+import java.util
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
+import scala.jdk.FutureConverters.CompletionStageOps
+import scala.reflect.ClassTag
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[couchbase] class CouchbaseCollectionSessionImpl(bucketSession: CouchbaseSession,
+                                                        scopeName: String,
+                                                        collectionName: String)
+    extends CouchbaseCollectionSession {
+
+  override def bucket: CouchbaseSession = bucketSession
+
+  override def scope: AsyncScope = bucket.underlying.scope(scopeName)
+  override def underlying: AsyncCollection = scope.collection(collectionName)
+
+  override def asJava = new CouchbaseCollectionSessionJavaAdapter(this)
+
+  override def insert[T: ClassTag](id: String, document: T): Future[Done] = {
+    underlying
+      .insert(id,
+              document,
+              InsertOptions
+                .insertOptions()
+                .transcoder(chooseTranscoder[T]()))
+      .asScala
+      .map(_ => Done)(ExecutionContext.parasitic)
+  }
+
+  override def insert[T: ClassTag](id: String, document: T, insertOptions: InsertOptions): Future[Done] = {
+    if (insertOptions.build.transcoder() == null) {
+      insertOptions.transcoder(chooseTranscoder[T]())
+    }
+    underlying
+      .insert(id, document, insertOptions)
+      .asScala
+      .map(_ => Done)(ExecutionContext.parasitic)
+  }
+
+  override def get[T: ClassTag](id: String): Future[CouchbaseDocument[T]] = {
+    val target: Class[T] = implicitly[ClassTag[T]].runtimeClass.asInstanceOf[Class[T]]
+    underlying
+      .get(id, GetOptions.getOptions.transcoder(chooseTranscoder[T]()))
+      .thenApply(gr => new CouchbaseDocument[T](id, gr.contentAs(target)))
```

Comment:
This will likely run on the common FJP if not already completed when `thenApply` is invoked (in which case it executes on the calling thread), better do it as a `map` on the Scala Future where execution context is explicit, to keep it on the same thread pool as the other stuff here.

Same for the other `thenApply`s in this class.